src/iterative/KOSMAPOSL/KOSMAPOSLReconstruction.cxx

```diff
@@ -0,0 +1,1162 @@
+    max_dz = (num_neighbours-1)/2;
+  }
+  min_dy = -(num_neighbours-1)/2;
+  max_dy = (num_neighbours-1)/2;
+  min_dx = -(num_neighbours-1)/2;
+  max_dx = (num_neighbours-1)/2;
+
+  Array<3,float> distance = Array<3,float>(IndexRange3D(min_dz, max_dz,
+                                                        min_dy, max_dy,
+                                                        min_dx, max_dx));
+
+  for (int z=min_dz;z<=max_dz;++z)
+    for (int y=min_dy;y<=max_dy;++y)
+      for (int x=min_dx;x<=max_dx;++x)
+        { // the distance is the euclidean distance:
+          //at the moment is used only for the definition of the neighbourhood
+
+          distance[z][y][x] =
+
+            sqrt(square(x*grid_spacing.x())+
+                 square(y*grid_spacing.y())+
+                 square(z*grid_spacing.z()));
+        }
+
+
+  int l=0,m=0, dimf_row=0;
+  int dimf_col = this->num_non_zero_feat-1;
+
+  dimf_row=this->num_voxels;
+
+  if(this->get_hybrid()){
+    calculate_norm_matrix (*this->kpnorm_sptr, dimf_row, dimf_col,
+                           current_alpha_estimate, distance);
+  }
+
+  //     calculate kernelised image
+
+  const int min_z = current_alpha_estimate.get_min_index();
+  const int max_z = current_alpha_estimate.get_max_index();
+
+  for (int z=min_z; z<=max_z; z++)
+    {
+      double pnkernel=0, kanatomical=0;
+      const int min_dz = max(distance.get_min_index(), min_z-z);
+      const int max_dz = min(distance.get_max_index(), max_z-z);
+
+      const int min_y = current_alpha_estimate[z].get_min_index();
+      const int max_y = current_alpha_estimate[z].get_max_index();
+
+
+      for (int y=min_y;y<= max_y;y++)
+        {
+          const int min_dy = max(distance[0].get_min_index(), min_y-y);
+          const int max_dy = min(distance[0].get_max_index(), max_y-y);
+
+          const int min_x = current_alpha_estimate[z][y].get_min_index();
+          const int max_x = current_alpha_estimate[z][y].get_max_index();
+
+
+          for (int x=min_x;x<= max_x;x++)
+            {
+              const int min_dx = max(distance[0][0].get_min_index(), min_x-x);
+              const int max_dx = min(distance[0][0].get_max_index(), max_x-x);
+
+              l = (z-min_z)*(max_x-min_x +1)*(max_y-min_y +1)
+                + (y-min_y)*(max_x-min_x +1)
+                + (x-min_x);
+
+              for (int dz=min_dz;dz<=max_dz;++dz)
+                for (int dy=min_dy;dy<=max_dy;++dy)
+                  for (int dx=min_dx;dx<=max_dx;++dx)
+                    {
+                      m = (dz-min_dz)*(max_dx-min_dx +1)*(max_dy-min_dy +1)
+                        + (dy-min_dy)*(max_dx-min_dx +1)
+                        + (dx-min_dx);
+
+                      if (get_hybrid()){
+                        if(current_alpha_estimate[z][y][x]==0){
+                          continue;
+                        }
+
+                        kPET
+                          = exp(-(*this->kpnorm_sptr)[0][l][m]
```

Comment:
Why is the first index always 0? This matrix seems never to be used elsewhere, why have this index?

Reply:
@danieldeidda

Reply:
this is a 2D matrix, I remember having problems when I used array<float,2>, so i made it a targetT